FIX control multi-threaded access to log file handle

pgrange opened the pull request:
We have several threads which might trace logs and, hence, write to the log file handle simultaneously. If don't pay attention to that we can end up with malformed, mixed logs.

So we introduce a mutex to control access to this file handle.

To check before merging:
* [x] CHANGELOG is up to date
* [x] Up to date with master

pgrange committed:
FIX control multi-threaded access to log file handle

We have several threads which might trace logs and, hence, write
to the log file handle simultaneously. If don't pay attention to
that we can end up with malformed, mixed logs.

So we introduce a mutex to control access to this file handle.

ffakenz reviewed (approved):
💯

ch1bo commented on hydra-node/src/Hydra/Logging.hs:
```diff
@@ -95,15 +96,13 @@ withTracerOutputTo ::
+      tracer =
+        Tracer $
+          mkEnvelope namespace >=> liftIO . write . Aeson.encode
+   in action tracer `finally` flushLogs
```
I don't see why this would be any different to before. I would expect `hPut` to write the whole bytestring to the handle atomically (because there is also a non-blocking variant).

ch1bo replied:
That sounds reasonable. I would suggest using `Aeson.encodeStrict` though.

pgrange replied:
@ch1bo 
>  I would expect hPut to write the whole bytestring to the handle atomically (because there is also a non-blocking variant).

Actually, hPut does documentation does not mention anything about atomicity. The non-blocking versus blocking versions of hPut are referring to more low-level _system-ish_ concerns.

The idea with the blocking version is to block the current thread until all the bytes have been written to the underlying buffers. That is, if, for any reason, the underlying buffers are full at some point while writing, the blocking version will wait for the buffer data to be flushed to some other lower level buffer/storage and then continue writing until all the data passed as a parameter have been written.

The non-blocking version will more write whatever it can to the underlying buffers and, as soon as something is full, instead of blocking, it will just return with the un-written data send back to the caller thread. In which case, the caller thread has to manage all the logic of re-submitting data later itself.

This is a big approximation, of course, but I think it's quite fair to describe the difference.

In neither of these two approaches is any atomicity or mutex involved. All the threads which call any of these functions on the same underlying file descriptor are in concurrence to access and write data to this file descriptor.

I think is it this concurrency which explains the mangling you've seen in the log outputs. I agree that you didn't see that much mangling, only _missing newlines_ but the problem with this un-determinism is that there is no way to really predict what will happen. Depending on how the Haskell thread scheduler works and depending on ho our application logs and its concurrency level, we might see very subtle/limited impact but, nevertheless, any code which would not put some sort of mutex to protect concurrent access to a file descriptor is unpredictable.

ch1bo reviewed (commented):
Why would you think this solves the problem we observed?

What i saw was multiple JSON messages in a single line. Not any more mangling. For example:
```
{"tag":"Event"}
{"tag":"Event"}{"tag":"Event"}
{"tag":"Event"}
```

abailly committed:
Use strict bytestrings when logging to stdout

There's normally no need for a mutex to serialise stdout writing because hPut is supposed
to be atomic already. However, I suspect that because we use lazy bytestrings and we
concatenate the newline, we end up having the writing of each _chunk_ to be atomic.

ch1bo reviewed (approved):
I do approve although I would prefer the `where` style over the `let` style :)

pgrange commented on hydra-node/src/Hydra/Logging.hs:
```diff
@@ -96,8 +96,7 @@ withTracerOutputTo ::
   (Tracer m msg -> IO a) ->
   IO a
 withTracerOutputTo hdl namespace action = do
-  hdlMutex <- newMVar ()
```
@abailly-iohk, we share a file descriptor between several concurrent threads so, without a mutex to protect this access, I'm afraid we can't ensure that byte strings are always written as we expect them to be.

Looking at the [hPut](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString.html#v:hPut) documentation I see no mention of any kind of guarantee when writing concurrently on shared file descriptor (Handle in Haskell lingo).

Of course, depending on how the Haskell thread system switching between threads, depending on how much laziness and logs frequencies we can have very diverse impact. That's the problem with undeterminism, sometimes it looks like it's working :)

But I wouldn't trust that code without this mutex, unless I've missed something obvious?

ch1bo replied:
You're not wrong.. I read hPut and it's constituents and it's just writing bytes to a file handle. No synchronization or thread-safety as far as I can tell. 

So why not go back to the original implementation?

Or instead of the empty mutex, just use a single element queue = MVar to sequentialize the writes like before and think how the to make the last write get flushed to avoid the race condition of #690?

pgrange replied:
@ch1bo, I think it's the simplest code that could work to, both, solve the original issue in #690 and solve the mangling problem you mentioned, as mutex implies less logic than queues.

pgrange commented:
Abandoned in favor of #711
